README.md
# JsonMatchers

Validate the JSON returned by your Rails JSON APIs

## Installation

Add this line to your application's `Gemfile`:

```ruby
group :test do
  gem "json_matchers"
end
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install json_matchers

## Usage

Inspired by [Validating JSON Schemas with an RSpec Matcher][original-blog-post].

[original-blog-post]: (https://robots.thoughtbot.com/validating-json-schemas-with-an-rspec-matcher)

First, configure it in your test suite's helper file:

### Configure

#### RSpec

`spec/spec_helper.rb`

```ruby
require "json_matchers/rspec"

JsonMatchers.schema_root = "spec/support/api/schemas"
```

#### Minitest

`test/test_helper.rb`

```ruby
require "minitest/autorun"
require "json_matchers/minitest/assertions"

JsonMatchers.schema_root = "test/support/api/schemas"

Minitest::Test.include(JsonMatchers::Minitest::Assertions)
```

### Declare

Declare your [JSON Schema](https://json-schema.org/learn/getting-started-step-by-step.html) in the schema
directory.

`spec/support/api/schemas/location.json` or
`test/support/api/schemas/location.json`:

Define your [JSON Schema](https://json-schema.org/learn/getting-started-step-by-step.html) in the schema
directory.

```json
{
  "id": "https://json-schema.org/geo",
  "$schema": "https://json-schema.org/draft-06/schema#",
  "description": "A geographical coordinate",
  "type": "object",
  "properties": {
    "latitude": {
      "type": "number"
    },
    "longitude": {
      "type": "number"
    }
  }
}
```

### Validate

#### RSpec

Validate a JSON response, a Hash, or a String against a JSON Schema with
`match_json_schema`:

`spec/requests/locations_spec.rb`

```ruby
describe "GET /locations" do
  it "returns Locations" do
    get locations_path, format: :json

    expect(response.status).to eq 200
    expect(response).to match_json_schema("locations")
  end
end
```

#### Minitest

Validate a JSON response, a Hash, or a String against a JSON Schema with
`assert_matches_json_schema`:

`test/integration/locations_test.rb`

```ruby
def test_GET_posts_returns_Locations
  get locations_path, format: :json

  assert_equal response.status, 200
  assert_matches_json_schema response, "locations"
end
```

### Embedding other Schemas

To re-use other schema definitions, include `$ref` keys that refer to their
definitions.

First, declare the singular version of your schema.

`spec/support/api/schemas/user.json`:

```json
{
  "id": "file:/user.json#",
  "type": "object",
  "required": ["id"],
  "properties": {
    "id": { "type": "integer" },
    "name": { "type": "string" },
    "address": { "type": "string" },
  },
}
```

Then, when you declare your collection schema, reference your singular schemas.

`spec/support/api/schemas/users/index.json`:

```json
{
  "id": "file:/users/index.json#",
  "type": "object",
  "definitions": {
    "users": {
      "description": "A collection of users",
      "example": [{ "id": "1" }],
      "type": "array",
      "items": {
        "$ref": "file:/user.json#"
      },
    },
  },
  "required": ["users"],
  "properties": {
    "users": {
      "$ref": "#/definitions/users"
    }
  },
}
```

NOTE: `$ref` resolves paths relative to the schema in question.

In this case `"user.json"` and `"users/index.json"` are resolved relative to
`"spec/support/api/schemas"` or `"test/support/api/schemas"`.

To learn more about `$ref`, check out
[Understanding JSON Schema Structuring][$ref].

[$ref]: https://spacetelescope.github.io/understanding-json-schema/structuring.html

### Declaring a schema in a Subdirectory

Nesting a schema within a subdirectory is also supported:

`spec/support/api/schemas/api/v1/location.json`:


```json
{
  "id": "https://json-schema.org/geo",
  "$schema": "https://json-schema.org/draft-06/schema#",
  "description": "A geographical coordinate",
  "type": "object",
  "properties": {
    "latitude": {
      "type": "number"
    },
    "longitude": {
      "type": "number"
    }
  }
}
```

`spec/requests/api/v1/locations_spec.rb`:

```ruby
describe "GET api/v1/locations" do
  it "returns Locations" do
    get locations_path, format: :json

    expect(response.status).to eq 200
    expect(response).to match_json_schema("api/v1/location")
  end
end
```

## Configuration

By default, the schema directory is `spec/support/api/schemas`.

This can be configured via `JsonMatchers.schema_root`.

```ruby
# spec/support/json_matchers.rb

JsonMatchers.schema_root = "docs/api/schemas"
```

## Upgrading from `0.9.x`

Calls to `match_json_schema` and `match_response_schema` no longer accept
options, and `JsonMatchers.configure` has been removed.

## Contributing

Please see [CONTRIBUTING].

`json_matchers` was inspired by [Validating JSON Schemas with an
RSpec Matcher][blog post] by Laila Winner.

`json_matchers` is maintained by Sean Doyle.

Many improvements and bugfixes were contributed by the [open source community].

[blog post]: https://robots.thoughtbot.com/validating-json-schemas-with-an-rspec-matcher
[CONTRIBUTING]: https://github.com/thoughtbot/json_matchers/blob/master/CONTRIBUTING.md
[open source community]: https://github.com/thoughtbot/json_matchers/graphs/contributors

## License

`json_matchers` is Copyright © 2018 thoughtbot.

It is free software, and may be redistributed under the terms specified in the
[LICENSE] file.

[LICENSE]: LICENSE.txt

## About thoughtbot

![thoughtbot](https://thoughtbot.com/logo.png)

`json_matchers` is maintained and funded by thoughtbot, inc.
The names and logos for thoughtbot are trademarks of thoughtbot, inc.

We love open source software!
See [our other projects][community].
We are [available for hire][hire].

[community]: https://thoughtbot.com/community?utm_source=github
[hire]: https://thoughtbot.com?utm_source=github


NEWS.md
master
======

0.11.1
======

* Follow first-level symlinks in the document store. [#99]

[#99]: https://github.com/thoughtbot/json_matchers/pull/99

0.11.0
======

* Expand references of DocuemntStore schemas ([#93], [3b66b4d])

[#93]: https://github.com/thoughtbot/json_matchers/pull/93
[3b66b4d]: https://github.com/thoughtbot/json_matchers/commit/3b66b4ddf369ec09b50cfe39c614266b34c7f3fe

0.10.0
======

* *BREAKING CHANGE* Implement the validation with the `json_schema` gem instead
  of `json-schema`. [#31]
* *BREAKING CHANGE* - remove support for configuring validation options.

[#31]: https://github.com/thoughtbot/json_matchers/pull/31

0.9.0
=====

* Deprecate support for configuring validation options.

0.8.0
=====

* Add `assert_json_matches_schema` and `refute_json_matches_schema` for
  use in `minitest` test suites [#66]

[#66]: https://github.com/thoughtbot/json_matchers/pull/66

0.7.3
=====

* Alias `match_response_schema` to `match_json_schema`. These matches can
  validate any JSON, whether it's from a request body, a response body, a Hash,
  or a String. [#68]

[#68]: https://github.com/thoughtbot/json_matchers/pull/68

0.7.2
=====

* Accept Array instances as arguments to `match_response_schema` [#60]

[#60]: https://github.com/thoughtbot/json_matchers/issues/60

0.7.1
=====

* Accept Hash instances as arguments to `match_response_schema` [#58], [#59]

[#58]: https://github.com/thoughtbot/json_matchers/pull/58
[#59]: https://github.com/thoughtbot/json_matchers/pull/59

0.7.0
=====

* Default `record_errors: true` [#52]
* Add support for configuring `record_errors` [#51]

[#52]: https://github.com/thoughtbot/json_matchers/pull/52
[#51]: https://github.com/thoughtbot/json_matchers/pull/51

0.6.3
=====

* Fix error message for string responses. [#49]

[#49]: https://github.com/thoughtbot/json_matchers/pull/49

0.6.2
=====

* Remove dependency on Rails. [#47]

[#47]: https://github.com/thoughtbot/json_matchers/pull/47

0.6.1
=====

* Configure default options for all matchers. [#46]
* Use `JSON.pretty_generate` to format error messages. [#44]

[#46]: https://github.com/thoughtbot/json_matchers/pull/46
[#44]: https://github.com/thoughtbot/json_matchers/pull/44

0.6.0
=====

* Accept a JSON string along with a `response`.

[#43]: https://github.com/thoughtbot/json_matchers/pull/43

0.5.1
=====

* Update `json-schema` dependency to `~> 2.6.0`. [#37]

[#37]: https://github.com/thoughtbot/json_matchers/pull/37

0.5.0
=====

Breaking Changes
----------------

* No longer default to `strict: true`.

0.4.0
=====

Breaking Changes
----------------

* Rename the gem to `json_matchers`.

0.3.1
=====

* No longer condone auto-loading RSpec. Add documentation around adding `require
  "json_matchers/rspec"` to consumers' `spec/spec_helper.rb`

Breaking Changes
----------------

* Rename module to `JsonMatchers`. This resolves clashing with
  gems like `oj` / `oj_mimic_json` that take control of the standard library's
  `json` module. As a result, the file to require is now `json_matchers`,
  instead of `json/matchers`.

0.3.0
=====

* Pass options from matcher to `JSON::Validator`

0.2.2
=====

* Includes validation failure message in RSpec output

0.2.1
=====

* Supports RSpec 2 syntax `failure_message_for_should` and
  `failure_message_for_should_not`

0.2.0
=====

* RSpec failure messages include both the body of the response, and the body of
  the JSON Schema file

0.1.0
=====

Breaking Changes
----------------

* Remove `schema_for` in favor of using `$ref`. To learn more about `$ref`,
  check out [Understanding JSON Schema Structuring](http://spacetelescope.github.io/understanding-json-schema/structuring.html)

0.0.1
=====

Features
--------

* Validate your Rails response JSON with `match_response_schema`


lib/json_matchers/rspec.rb
require "json_matchers"
require "json_matchers/assertion"

module JsonMatchers
  self.schema_root = File.join("spec", "support", "api", "schemas")
end

RSpec::Matchers.define :match_json_schema do |schema_name, **options|
  if options.present?
    warn <<-WARN
DEPRECATION:

  After `json_matchers@0.9.x`, calls to `match_json_schema` and
  `match_response_schema` will no longer accept options.

  See https://github.com/thoughtbot/json_matchers/pull/31 for more information.

WARN
  end

  assertion = JsonMatchers::Assertion.new(schema_name.to_s, options)

  match do |json|
    assertion.valid?(json)
  end

  if respond_to?(:failure_message)
    failure_message do
      assertion.valid_failure_message
    end

    failure_message_when_negated do
      assertion.invalid_failure_message
    end
  else
    failure_message_for_should do
      assertion.valid_failure_message
    end

    failure_message_for_should_not do
      assertion.invalid_failure_message
    end
  end
end

RSpec::Matchers.alias_matcher :match_response_schema, :match_json_schema
